apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricRegistry.java

```diff
@@ -0,0 +1,121 @@
+     *               Note that there will be a {@link MetricSet} created for each distinct set of tags.
+     * @param metric this supplier will be called for every reporting cycle
+     *               ({@link co.elastic.apm.agent.report.ReporterConfiguration#metricsInterval metrics_interval)})
+     * @see #add(String, Map, DoubleSupplier)
+     */
+    public void addUnlessNan(String name, Map<String, String> tags, DoubleSupplier metric) {
+        if (!Double.isNaN(metric.get())) {
+            add(name, tags, metric);
+        }
+    }
+
+    /**
+     * Same as {@link #add(String, Map, DoubleSupplier)} but only adds the metric
+     * if the {@link DoubleSupplier} returns a positive number or zero.
+     *
+     * @param name   the name of the metric
+     * @param tags   tags for the metric.
+     *               Tags can be used to create different graphs based for each value of a specific tag name, using a terms aggregation.
+     *               Note that there will be a {@link MetricSet} created for each distinct set of tags.
+     * @param metric this supplier will be called for every reporting cycle
+     *               ({@link co.elastic.apm.agent.report.ReporterConfiguration#metricsInterval metrics_interval)})
+     * @see #add(String, Map, DoubleSupplier)
+     */
+    public void addUnlessNegative(String name, Map<String, String> tags, DoubleSupplier metric) {
+        if (metric.get() >= 0) {
+            add(name, tags, metric);
+        }
+    }
+
+    /**
+     * Adds a gauge to the metric registry.
+     *
+     * @param name   the name of the metric
+     * @param tags   tags for the metric.
+     *               Tags can be used to create different graphs based for each value of a specific tag name, using a terms aggregation.
+     *               Note that there will be a {@link MetricSet} created for each distinct set of tags.
+     * @param metric this supplier will be called for every reporting cycle
+     *               ({@link co.elastic.apm.agent.report.ReporterConfiguration#metricsInterval metrics_interval)})
+     */
+    public void add(String name, Map<String, String> tags, DoubleSupplier metric) {
+        MetricSet metricSet = metricSets.get(tags);
+        if (metricSet == null) {
+            metricSets.putIfAbsent(tags, new MetricSet(tags));
+            metricSet = metricSets.get(tags);
+        }
+        metricSet.add(name, metric);
+    }
+
+    public void serialize(JsonWriter jw, StringBuilder replaceBuilder) {
+        MetricRegistrySerializer.serialize(this, replaceBuilder, jw);
+    }
+
+    public double get(String name, Map<String, String> tags) {
+        final MetricSet metricSet = metricSets.get(tags);
+        if (metricSet != null) {
+            return metricSet.get(name).get();
+        }
+        return Double.NaN;
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Moving the serialization code to the `serialize` package is great, so I think having this `JsonWriter` residual here is a shame. Sorry for nagging, I have future refactors in mind...
What do you say adding a static `asString(MetricsRegistry)` to the serializer?

Reply:
do you want to create a separate module for serialization?

Reply:
Not at the moment :)
I just think it makes sense to avoid dependency on serialization lib here.

Reply:
I'm all about keeping package dependencies uni-directional!